File: dac/backend/src/main/java/com/dremio/dac/model/sources/Source.java
/*
 * Copyright (C) 2017 Dremio Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dremio.dac.model.sources;

import com.dremio.service.namespace.source.proto.SourceType;
import com.fasterxml.jackson.annotation.JsonIgnore;

import io.protostuff.ByteString;
import io.protostuff.LinkedBuffer;
import io.protostuff.ProtobufIOUtil;
import io.protostuff.Schema;

/**
 * Source model.
 */
public abstract class Source {
  public static <S extends Source> S fromByteString(SourceType type, ByteString config) {
    @SuppressWarnings("unchecked")
    final S source = (S) fromSource(SourceDefinitions.getSourceSchema(type), config);
    return source;
  }

  @JsonIgnore
  public SourceType getSourceType() {
    // Cannot be done at construction time as:
    // - SourceDefinitions stores Source classes
    // - Protostuff classes have a static DEFAULT_INSTANCE field
    // - it's an initialization loop...
    return SourceDefinitions.getType(this.getClass());
  }

  public ByteString toByteString() {
    @SuppressWarnings("unchecked")
    Schema<Source> schema = (Schema<Source>) SourceDefinitions.getSourceSchema(getSourceType());

    LinkedBuffer buffer = LinkedBuffer.allocate(512);
    byte[] bytes = ProtobufIOUtil.toByteArray(this, schema, buffer);

    return ByteString.copyFrom(bytes);
  }

  private static <S extends Source> S fromSource(Schema<S> schema, ByteString config) {
    final S source = schema.newMessage();
    if (config != null) {
      ProtobufIOUtil.mergeFrom(config.toByteArray(), source, schema);
    }
    return source;
  }
}

File: ui/dremio-js/src/interfaces/Source.ts
/*
 * Copyright (C) 2017-2019 Dremio Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { CatalogObjectMethods } from "./CatalogObject.js";
import type { CatalogReference } from "./CatalogReference.js";

export type CommunitySourceProperties = {
  acceleration: {
    activePolicyType: string;
    gracePeriod: Temporal.Duration;
    neverExpire: boolean;
    neverRefresh: boolean;
    refreshPeriod: Temporal.Duration;
    refreshSchedule: string;
  };

  allowCrossSourceSelection: boolean;
  config: {
    accessKey: string;
    compatibilityMode: boolean;
    credentialType: string;
    defaultCtasFormat: string;
    enableAsync: boolean;
    enableFileStatusCheck: boolean;
    externalBucketList: string[];
    isCachingEnabled: boolean;
    isPartitionInferenceEnabled: boolean;
    maxCacheSpacePct: number;
    requesterPays: boolean;
    rootPath: string;
    secure: boolean;
  };
  createdAt: Date;
  disableMetadataValidityCheck: boolean;
  id: string;
  metadataPolicy: {
    authTTL: Temporal.Duration;
    autoPromoteDatasets: boolean;
    datasetExpireAfter: Temporal.Duration;
    datasetRefreshAfter: Temporal.Duration;
    datasetUpdateMode: string;
    deleteUnavailableDatasets: boolean;
    namesRefresh: Temporal.Duration;
  };
  name: string;
  status: string;
  type: string;
};

export type CommunitySourceMethods = {
  children(): {
    data(): AsyncGenerator<CatalogReference>;
  };
} & CatalogObjectMethods;

export type CommunitySource = CommunitySourceProperties &
  CommunitySourceMethods;
